### ISSUE TITLE
Set up cron job for adding Wikidata ID statements to MaRDI via DOI

### ISSUE THREAD

**Issue by Daniel-Mietchen** [MaRDI4NFDI/portal-compose#580]
**New feature description in words**:

There is considerable overlap between MaRDI and Wikidata, and with ongoing curation on both ends, it makes sense to set up some tooling to assist the syncing. Let's test these waters with a cron job for adding Wikidata ID statements to MaRDI via DOIs known to both.

**TODOS**: 
-  query MaRDI for items that have a DOI known to Wikidata but no Wikidata ID (i.e. MaRDI P12) statement
-  convert query results into edits. [This query](https://query.portal.mardi4nfdi.de/#%23%20This%20query%20produces%20statements%20for%0A%23%20https%3A%2F%2Fquickstatements.portal.mardi4nfdi.de%2F%23%2Fbatch%2Fnew%0A%23%20Help%3A%20https%3A%2F%2Fwww.wikidata.org%2Fwiki%2FHelp%3AQuickStatements%20.%0A%0APREFIX%20wdt%3A%20%3Chttps%3A%2F%2Fportal.mardi4nfdi.de%2Fprop%2Fdirect%2F%3E%0APREFIX%20wd%3A%20%3Chttps%3A%2F%2Fportal.mardi4nfdi.de%2Fentity%2F%3E%0A%0APREFIX%20wikidata_wd%3A%20%3Chttp%3A%2F%2Fwww.wikidata.org%2Fentity%2F%3E%0APREFIX%20wikidata_wdt%3A%20%3Chttp%3A%2F%2Fwww.wikidata.org%2Fprop%2Fdirect%2F%3E%0A%0ASELECT%20DISTINCT%0A%0A%23%23%23%23%23%23%23%0A%23%20To%20get%20the%20QuickStatements%20commands%2C%20comment%20out%20the%20lines%20until%20the%20next%20hash%20line%0A%23%20%3Fmardi_item%0A%23%20%3Fmardi_item_label%0A%23%20%3Fwikidata%0A%23%20%3Fwikidata_label%0A%23%20%3FDOI%0A%23%23%23%23%23%23%23%0A%0A%28REPLACE%28STR%28%3Fmardi_item%29%2C%20%22.%2aQ%22%2C%20%22Q%22%29%20AS%20%3Fmardi_qid%29%20%0A%28%22P12%22%20AS%20%3Fproperty%29%0A%28CONCAT%28%22%5C%22%22%2C%20REPLACE%28STR%28%3Fwikidata%29%2C%20%22.%2aQ%22%2C%20%22Q%22%29%2C%20%22%5C%22%22%29%20AS%20%3Fwikidata_qid%29%20%0A%0AWHERE%20%7B%0A%20%20SERVICE%20bd%3Asample%20%7B%20%3Fmardi_item%20wdt%3AP27%20%3Fdoi%20.%20bd%3AserviceParam%20bd%3Asample.limit%2010000%20%7D%0A%20%20BIND%28UCASE%28%3Fdoi%29%20AS%20%3FDOI%29%0A%20%20FILTER%20NOT%20EXISTS%20%7B%20%3Fmardi_item%20wdt%3AP12%20%3FWikidataQID%20%7D%20%0A%20%20service%20%3Chttps%3A%2F%2Fquery.wikidata.org%2Fsparql%3E%20%7B%0A%20%20%20%20%3Fwikidata%20wikidata_wdt%3AP356%20%3FDOI%20.%0A%20%20%20%20%3Fwikidata%20rdfs%3Alabel%20%3Fwikidata_label%20.%0A%20%20%20%20FILTER%20%28LANG%28%3Fwikidata_label%29%20%3D%20%22en%22%29%0A%20%20%7D%0A%20%20%3Fmardi_item%20rdfs%3Alabel%20%3Fmardi_item_label%20.%0A%20%20FILTER%20%28LANG%28%3Fmardi_item_label%29%20%3D%20%22en%22%29%0A%7D%0A%23%20LIMIT%201000) yields corresponding QuickStatements commands, but a bot job might be preferable.
- do some quality checks (e.g. whether the respective DOIs resolve) and log the results

**Considered Alternatives**: 
- The sample query above uses the Wikidata endpoint for the full graph. In the long term, we should probably replace that with the one for the scholarly graph.
- Instead of SPARQL, we might want to use the REST API, which should be faster and not time out so easily.

**Checklist for this issue**: 
- [x] Participants in discussion have been invited as assignees 
- [x] All fields of the issue have been filled 
- [x] Example fields have been removed
- [x] The main MaRDI project has been assigned to this issue

** Progress **
- [x] Optimize [query](https://tinyurl.com/2bg4gm34) for php 
```SPARQL
PREFIX wdt: <https://portal.mardi4nfdi.de/prop/direct/>
PREFIX wikidata_wdt: <http://www.wikidata.org/prop/direct/>

SELECT DISTINCT
(REPLACE(STR(?mardi_item), ".*Q", "Q") AS ?mardi_qid) 
(REPLACE(STR(?wikidata), ".*Q", "Q") AS ?wikidata_qid) 
WHERE {
  SERVICE bd:sample { ?mardi_item wdt:P27 ?doi . bd:serviceParam bd:sample.limit 10000 }
  BIND(UCASE(?doi) AS ?DOI)
  FILTER NOT EXISTS { ?mardi_item wdt:P12 ?WikidataQID } 
  service <https://query.wikidata.org/sparql> {
    ?wikidata wikidata_wdt:P356 ?DOI 
  }
}
```
- [ ] Set up one-time script
- [ ] Check edits

**Comment by eloiferrer** [comment]
Since most of this overlap concerns items that we have imported from zbMATH, I think a cleaner workflow would be to add the corresponing wikidata QID as part of the preprocessing of the zbMATH data, before it is inserted to the graph through the zbmath-importer. Does that make sense or would that be feasible @LizzAlice ?

**Comment by LizzAlice** [comment]
We can do that, but for the majority of the items, we will have to make a bot job anyway, since there are a lot of existing zbmath articles. So maybe it would more sense to just keep the separate bot job so that we can also catch items that are not from zbmath

**Comment by physikerwelt** [comment]
Before I can start implementing that job, I would need to know the details about

- do some quality checks (e.g. whether the respective DOIs resolve) and log the results

How does that relate to the other task?

**Comment by Daniel-Mietchen** [comment]
> Since most of this overlap concerns items that we have imported from zbMATH, I think a cleaner workflow would be to add the corresponing wikidata QID as part of the preprocessing of the zbMATH data, before it is inserted to the graph through the zbmath-importer. Does that make sense or would that be feasible @LizzAlice ?

One issue with that is that the Wikidata entry for a paper might well be created after the corresponding entry in the MaRDI portal.

**Comment by Daniel-Mietchen** [comment]
@physikerwelt 

> Before I can start implementing that job, I would need to know the details about
> 
>     * do some quality checks (e.g. whether the respective DOIs resolve) and log the results
> 
> 
> How does that relate to the other task?

I see two main quality checks here:
1. check that URLs of the format ```https://doi.org/$1``` resolve, perhaps with occasional checks that the ```http://doi.org/$1```, ```https://dx.doi.org/$1``` and ```http://dx.doi.org/$1``` variants do too (usually, if one of the four works, the others do too, but there are edge cases)
1. check that the MaRDI version of the DOI is properly case-normalized, as per #594 .

In terms of logging the results, I think this is a nice to have rather than essential feature. However, given that DOIs are widely presumed to be "persistent" identifiers, while that's not always true in practice, some actual data on the matter would be nice.
I think it would be best to keep track of
- the earliest date at which we confirmed a given DOI to resolve
- the latest date at which we confirmed a given DOI to resolve
- the earliest date at which we confirmed a given DOI to not resolve
- the latest date at which we confirmed a given DOI to not resolve

Intermediate results between these dates can usually be deleted from the logs if the results of consecutive checks are identical. Important edge cases are when the periods overlap during which DOIs resolve or don't. Here, we should probably err on the side of keeping until we've had a closer look at the patterns in the data.

As for the relation to 
- #581 ,

I think two things should be kept in mind:
- it's probably best to run the scripts for #580 and #581 (or similar ones that we may have in the future) after each other, not in parallel
- both make use of DOIs to identify "the same" entity to be mapped between the portal and Wikidata, which works better if the DOIs are normalized, as per #594.

**Comment by physikerwelt** [comment]
I think the doi checks should be a task on its own.
There are several script for doing this already see for example

https://github.com/altmetric/php-identifiers
https://github.com/continuous-foundation/doi-utils
...

**Comment by physikerwelt** [comment]
@Daniel-Mietchen please review https://portal.mardi4nfdi.de/wiki/Special:Contributions/Import241212081232

**Comment by physikerwelt** [comment]
The query now leads to empty results. @Daniel-Mietchen please reopen, if there is something wrong.
